src/compiler/checker.ts

```diff
@@ -17592,6 +17592,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getTypeOfSymbol(symbol); // intentionally doesn't use resolved symbol so type is cached as expected on the alias
         }
         else {
+            tryGetDeclaredTypeOfSymbol(resolvedSymbol); // call this first to ensure typeParameters is populated (if applicable)

```

Comment:
is there a reason you can't inspect the type returned from `tryGetDeclaredTypeOfSymbol` for type parameters? Or are `TypeParameter`s only available only symbol(links) instead of type references?

Reply:
@sandersn

> is there a reason you can't inspect the type returned from tryGetDeclaredTypeOfSymbol for type parameters? Or are TypeParameters only available only on Symbol(Links) instead of type references?

So I tried the approach that you suggested and I observed that the `type` returned by `tryGetDeclaredTypeOfSymbol()` is not `InterfaceType` but `TypeReference`, where `target` is `InterfaceType`. When I try `target.typeParameters`, this unfortunately doesn't work because the elements of `target.typeParameters` do not have `constraint` property for some reason.

Next, I examined other uses of `checkTypeArgumentConstraint()` and it seems like `typeParameters` are obtained via `getTypeParametersForTypeReference()`. When you check the implementation of that function, it seems to be conceptually equivalent to what I was doing (it uses `getSymbolLinks()` at the end of the day):
```typescript
    function getTypeParametersForTypeReference(node: TypeReferenceNode | ExpressionWithTypeArguments) {
        const type = getTypeFromTypeReference(node);
        if (!isErrorType(type)) {
            const symbol = getNodeLinks(node).resolvedSymbol;
            if (symbol) {
                return symbol.flags & SymbolFlags.TypeAlias && getSymbolLinks(symbol).typeParameters ||
                    (getObjectFlags(type) & ObjectFlags.Reference ? (type as TypeReference).target.localTypeParameters : undefined);
            }
        }
        return undefined;
    }
```
So I ended up extracting  a `getTypeParametersForTypeAndSymbol()` sub-routine in order to leverage the same logic for handling edge cases. Does that make more sense to you now?

Reply:
Yes, extracting the existing code seems like a good solution.